src/libraries/System.Security.Cryptography/tests/X509Certificates/X509FilesystemTests.Windows.cs

```diff
@@ -0,0 +1,464 @@
+                        {
+                            keyBag.Attributes.Add(s_machineKey);
+                        }
+
+                        certs.AddCertificate(cert);
+                    }
+                }
+            }
+
+            for (int i = 0; i < certOnlyCount; i++)
+            {
+                using (RSA key = RSA.Create(1024))
+                {
+                    CertificateRequest req = new CertificateRequest(
+                        $"CN={name}.co.{i}",
+                        key,
+                        HashAlgorithmName.SHA256,
+                        RSASignaturePadding.Pkcs1);
+
+                    using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
+                    {
+                        certs.AddCertificate(cert);
+                    }
+                }
+            }
+
+            Pkcs12Builder builder = new Pkcs12Builder();
+            builder.AddSafeContentsEncrypted(certs, "", pbeParams);
+            builder.AddSafeContentsUnencrypted(keys);
+            builder.SealWithMac("", HashAlgorithmName.SHA1, 1);
+            return builder.Encode();
+        }
+
+        private static void EnsureNoKeysGained<TState>(TState state, Func<TState, IDisposable> importer)
+        {
+            const int ERROR_ACCESS_DENIED = (unchecked((int)0x80010005));
+
+            // In the good old days, before we had threads or parallel processes, these tests would be easy:
+            // * Read the directory listing(s)
+            // * Import a thing
+            // * See what new things were added
+            // * Dispose the thing
+            // * See that the new things went away
+            //
+            // But, since files can be created by tests on other threads, or even by other processes,
+            // recheck the directory a few times (MicroRetryCount) after sleeping (SleepMs).
+            //
+            // Sadly, that's not sufficient, because an extra file gained during that window could itself
+            // be leaked, or be intentionally persisted beyond the recheck interval.  So, instead of failing,
+            // try again from the beginning.  If we get parallel leaked on MacroRetryCount times in a row
+            // we'll still false-fail, but unless a majority of the tests in the process are leaking keys,
+            // it's unlikely.
+            //
+            // Before changing these constants to bigger numbers, consider the combinatorics. Failure will
+            // sleep (MacroRetryCount * (MicroRetryCount - 1) * SleepMs) ms, and also involves non-zero work.
+            // Failing 29 tests at (3, 5, 1000) adds about 6 minutes to the test run compared to success.
+
+            const int MacroRetryCount = 3;
+            const int MicroRetryCount = 5;
+            const int SleepMs = 1000;
+
+            KeyPaths keyPaths = KeyPaths.GetKeyPaths();
+            HashSet<string> gainedFiles = null!;
```

Comment:
I don't think unit test projects are nullable aware, so the bang is not needed here.